geps/gep-3875/index.md

```diff
@@ -0,0 +1,192 @@
+
+## Non-Goals
+
+(What is out of scope for this proposal.)
+
+## Introduction
+
+(Can link to external doc -- but we should bias towards copying
+the content into the GEP as online documents are easier to lose
+-- e.g. owner messes up the permissions, accidental deletion)
+
+## API
+
+This proposal includes a few API changes:
+
+1. A new `namespaces` field will be added to `spec.targetRefs`.
+2. A new `from` field will be added to `spec.targetRefs`.
+3. A new `mode` field will be added to `spec`.
+
+### `TargetRef.Namespace`
+
+Implementation target: Experimental in 1.4.
+
+The `namespace` field allows a service consumer in another namespace to override the BackendTLSPolicy to use when connecting to a Service.
+
+* `BackendTLSPolicy` can be applied to a `Service` in the same namespace to provide defaults to clients.
+* `BackendTLSPolicy` can be applied to a `Service` in a different namespace to provide explicit configuration for clients in that namespace.
+
+In a complex setup, there are up to 4 namespaces involved: the Gateway, ListenerSet, Route, and Service.
+However, `BackendTLSPolicy` will only be consulted in the `Gateway` and `Service` namespace.
+For mesh use cases, there are only 2 relevant namespaces: the calling client namespace and the `Service` namespace; both of these will be consulted.
+
+If a `BackendTLSPolicy` exists for a `Service` in both namespaces, the consumer namespace will take precedence.
+
+For example, given the following (condensed) configuration:
+
+```yaml
+kind: Gateway
+namespace: gateway-ns
+---
+kind: BackendTLSPolicy
+name: gateway-pol
+namespace: gateway-ns
+targetRef:
+  name: svc
+  namespace: app
+---
+kind: Service
+name: svc
+namespace: app
+---
+kind: BackendTLSPolicy
+name: app-pol
+namespace: gateway-ns
+targetRef:
+  name: svc
+```
+
+For requests from Gateways in the `gateway-ns` to `svc.app`, the `gateway-pol` will be used. For requests from Gateways in other namespaces, the `app-pol` will be used.
+
+There will be no merging of policies. Exactly one policy will apply.
+This mirrors the [producer and consumer route semantics](https://gateway-api.sigs.k8s.io/concepts/glossary/?h=gloss#producer-route).
+
+This will be implemented by a new API type, `PolicyTargetReferenceWithSectionName`, as there is no existing policy that utilizes a cross-namespace reference.
+
+### `TargetRef.From`
+
+This field is expected to be implemented in a release after [`Namespace`](#targetrefnamespace).
+
+[Tracking Issue](https://github.com/kubernetes-sigs/gateway-api/issues/3856)
+
+While `Namespace` scoping enables some override use cases, the granularity is coarse.
+For example, we cannot apply different policies when requests come from the same namespace.
+It is (unfortunately) common to run many applications in a single namespace, or even to want different policies from a single workload.
+
+This proposal adds a new `from` selector on the `targetRef`, which would allow referencing an object in the same namespace:
```

Comment:
This likely requires a corresponding update to our policy attachment spec.

/cc @youngnick @guicassolato

Reply:
Hmm, I guess a question on this. Does this change mean that all 'policies' should/must also do this?

Or will it be isolated to TLS (at least for now, and the policy attachment spec will just list it as _an option_ for policies?)

Reply:
Given how complex the precedence is here (Route > ListenerSet > Gateway > `from` unset, in the consumer namespace > `from` unset, in the producer namespace), I'd _really_ like to have some tooling that could tell a user what the effective policy is. Today that means gwctl. One of the constraints of that is that it can only implement the policy spec and not have special cases per policy.

Reply:
I agree that in any reasonably complex deployment, it will be nigh impossible for a human to understand which settings are in effect, and tooling will be basically _required_.

For upstream Policy objects, it's important that every single thing we do is a thing that we have written down a pattern for - as consumers of Gateway API will, with good reason, use our design decisions as a basis and justification for their own. At a minimum, we will need to update Policy Attachment to define how `from` works, and how it interacts with other Policy dimensions like dynamic merge strategies, which also add complexity.

I'm not saying we _can't_ do this, but I think it's important to spend some time thinking about the most complex use cases and how bad the complexity will be in them.